crates/hc_seed_bundle/src/unlocked_seed_bundle.rs

```diff
@@ -1,3 +1,139 @@
+    /// LockedSeedCiphers to be used for decrypting the bundle.
+    pub async fn from_locked(
+        bytes: &[u8],
+    ) -> SodokenResult<Vec<crate::LockedSeedCipher>> {
+        crate::LockedSeedCipher::from_locked(bytes)
+    }
+
+    /// Derive a new sub SeedBundle by given index.
+    pub fn derive(
+        &self,
+        index: u32,
+    ) -> impl Future<Output = SodokenResult<Self>> + 'static + Send {
+        let seed = self.seed.clone();
+        async move {
+            let new_seed = sodoken::BufWriteSized::new_mem_locked()?;
+            sodoken::kdf::derive_from_key(
+                new_seed.clone(),
+                index as u64,
+                *KDF_CONTEXT,
+                seed,
+            )?;
+            Self::priv_from_seed(new_seed.to_read_sized()).await
+        }
+    }
+
+    /// Get the signature pub key generated by this seed.
+    pub fn get_sign_pub_key(
+        &self,
+    ) -> sodoken::BufReadSized<{ sodoken::sign::SIGN_PUBLICKEYBYTES }> {
+        self.sign_pub_key.clone()
+    }
+
+    /// Sign some data with the secret key generated by this seed.
+    pub fn sign_detached<M>(
+        &self,
+        message: M,
+    ) -> impl Future<
+        Output = SodokenResult<
+            sodoken::BufReadSized<{ sodoken::sign::SIGN_BYTES }>,
+        >,
+    >
+           + 'static
+           + Send
+    where
+        M: Into<sodoken::BufRead> + 'static + Send,
+    {
+        let sign_sec_key = self.sign_sec_key.clone();
+        async move {
+            let sig = sodoken::BufWriteSized::new_no_lock();
+            sodoken::sign::sign_detached(sig.clone(), message, sign_sec_key)
+                .await?;
+            Ok(sig.to_read_sized())
+        }
+    }
+
+    /// Get the raw appData bytes.
+    pub fn get_app_data_bytes(&self) -> &[u8] {
+        &self.app_data
+    }
+
+    /// Set the raw appData bytes.
+    pub fn set_app_data_bytes<B>(&mut self, app_data: B)
+    where
+        B: Into<Arc<[u8]>>,
+    {
+        self.app_data = app_data.into();
+    }
+
+    /// Get the decoded appData bytes by type.
+    pub fn get_app_data<T>(&self) -> SodokenResult<T>
+    where
+        T: 'static + for<'de> serde::Deserialize<'de>,
+    {
+        rmp_serde::from_read_ref(&self.app_data).map_err(SodokenError::other)
+    }
+
+    /// Set the encoded appData bytes by type.
+    pub fn set_app_data<T>(&mut self, t: &T) -> SodokenResult<()>
+    where
+        T: serde::Serialize,
+    {
+        let mut se = rmp_serde::encode::Serializer::new(Vec::new())
+            .with_struct_map()
+            .with_string_variants();
+        t.serialize(&mut se).map_err(SodokenError::other)?;
+        self.app_data = se.into_inner().into_boxed_slice().into();
+        Ok(())
+    }
+
+    /// Get a SeedCipherBuilder that will allow us to lock this bundle.
+    pub fn lock(&self) -> crate::SeedCipherBuilder {
+        crate::SeedCipherBuilder::new(self.seed.clone(), self.app_data.clone())
+    }
```

Comment:
I wonder if it would be more appropriate API-wise to have this consume `self`? "Locking" seems like it should be a destructive act that makes the unlocked thing inaccessible.

Reply:
Mmm, what you're saying makes sense. But I feel like the ergonomics of the lib would be damaged in that case, and there's no security reason not to keep the unlocked struct as well... Which leaves changing the name to something else. I think I'll leave it as-is for now, but feel free to suggest something else : )